include/kodlab_mjbots_sdk/behavior_manager.h

```diff
@@ -0,0 +1,295 @@
+   * @brief Behavior command that kills robot, default -1
+   */
+  static const int KILL_ROBOT_IDX = -1;
+
+  /**
+   * @brief Behavior command that turns robot off, always 0
+   */
+  static const int DEFAULT_IDX = 0;
+
+  /**
+   * @brief Behaviors available to the robot
+   * @note The first element of this vector is always `OFF_BEHAVIOR`.  Added
+   *       behaviors are appended to the end.
+   */
+  std::vector<std::unique_ptr<Behavior<Robot>>> behaviors_;
+
+  /**
+   * @brief Names of behaviors in `behaviors_`
+   */
+  std::vector<std::string> names_;
+
+  /**
+   * @brief Index in `behaviors_` of the selected behavior
+   */
+  unsigned int selected_idx = 0;
+
+  /**
+   * @brief Index in `behaviors_` of the incoming behavior while transitioning
+   *        behaviors
+   */
+  int next_idx_ = 0;
+
+  /**
+   * @brief Indicates whether a behavior switch is in progress
+   */
+  bool switching_behaviors_ = false;
+
+public:
+  /**
+   * @brief Construct a behavior manager object
+   * @param robot pointer to robot that the behavior is executing on
+   */
+  BehaviorManager(std::shared_ptr<Robot> robot) : robot_(robot)
+  {
+    AddBehavior<OffBehavior<Robot>>(robot_, "OFF");
+  }
+
+  /**
+   * @brief Destructor
+   */
+  virtual ~BehaviorManager() {}
+
+  /**
+   * @brief Set the current behavior
+   * @param next_idx index in `behaviors_`
+   */
+  virtual void SetBehavior(const int &next_idx)
+  {
+    if (next_idx == KILL_ROBOT_IDX)
+    {
+      kodlab::CTRL_C_DETECTED = true; // kill robot
+      LOG_FATAL("Behavior %d received, killing robot.", KILL_ROBOT_IDX);
+    } else if (next_idx == selected_idx)
+    {
+      LOG_INFO("Behavior %d is already active.", next_idx);
+      return;
+    }
+    else if (next_idx < behaviors_.size() && next_idx >= 0)
+    {
+      next_idx_ = next_idx; // update next behavior
+      if (!behaviors_[next_idx_]->is_initialized())
+      {
+        behaviors_[next_idx_]->Init();
+      }
+      behaviors_[selected_idx]->Stop(*behaviors_[next_idx_]); // stop current behavior
+      switching_behaviors_ = true;  // indicate transition is in progress
+    } else
+    {
+      LOG_WARN("Invalid behavior selected: %d", next_idx);
+    }
+  }
+
+  /**
+   * @brief Constructs and initializes a behavior
+   * @tparam BehaviorType `kodlab::Behavior`-derived type
+   * @tparam ConstructorArgs constructor arguments
+   * @param args behavior constructor arguments
+   * @return unique pointer to the behavior if initialization was successful,
+   *         nullptr otherwise
+   */
+  template<class BehaviorType, typename... ConstructorArgs>
+  std::unique_ptr<BehaviorType> InitBehavior(ConstructorArgs &&... args)
+  {
+    static_assert(std::is_base_of<kodlab::Behavior<Robot>, BehaviorType>::value,
+                  "BehaviorType must be a `kodlab::Behavior`-derived type");
+    auto b = std::make_unique<BehaviorType>(args...);
+    if (b->Init())
+    {
+      b->set_initialized();
+      return std::move(b);
```

Comment:
What is the purpose of this function? When do you expect it to be used?

Reply:
see above comment

Reply:
This function constructs the behavior from the given arguments and calls the behavior's initialization method.  If the initialization method succeeds, the behavior is returned to be added to the behaviors list.  Otherwise, a `nullptr` is returned.

The name change and move to private sound good!

Reply:
Changed in d5a6010ee947e62b6bbd122659ed3fcc02ff3224